In [1]:
import pandas as pd
import numpy as np

Possible Analysis Points

- Sellers with highest reviews scores/ popular seller prices
- Reviews score per region
- Delivery time per region
- average items bought per category at a time
- 
- Review for category for product
- How many customer pay in installments



In [2]:
#Datasets
customerData = pd.read_csv("olist_customers_dataset.csv")
geolocation = pd.read_csv("olist_geolocation_dataset.csv")
payments = pd.read_csv("olist_order_payments_dataset.csv")

products = pd.read_csv("olist_products_dataset.csv")
prod_cat = pd.read_csv("product_category_name_translation.csv")

order_reviews= pd.read_csv("olist_order_reviews_dataset 2.csv")
order_items = pd.read_csv("olist_order_items_dataset.csv")
orders = pd.read_csv("olist_orders_dataset.csv")

sellers = pd.read_csv("olist_sellers_dataset.csv")






<span style='font-size:large'>CLEANED DATA</span>

- <span style='font-size:medium'>products\_clean</span>
- <span style='font-size:large'>order\_items\_clean</span>
- <span style='font-size:medium'>order\_revieworder\_res\_clean2</span>

<span style='font-size:x-large'>ALREADY CLEAN</span>

- Customers
- geolocation
- payments
- Sellers
- Orders



In [3]:
print(len(customerData))
customer_count= customerData['customer_unique_id'].value_counts()

customer_count

99441


8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
ca77025e7201e3b30c44b472ff346268     7
6469f99c1f9dfae7733b25662e7f1782     7
                                    ..
081f07439678af2da7755a0aa572154d     1
9c57789a6a587f4efcfb01e36c106014     1
9e4c21635f640562de1a6976374fba03     1
06c0ea6ee892364d1608ee47aa9f56a0     1
84732c5050c01db9b23e19ba39899398     1
Name: customer_unique_id, Length: 96096, dtype: int64

In [21]:


# print(len(geolocation))
geolocation_new = geolocation.replace(['são paulo'],['sao paulo'])
cities=geolocation_new['geolocation_city'].value_counts()
saoPaulo = geolocation[(geolocation_new == 'sao paulo').any(axis=1)]
print(cities)

sao paulo               160718
rio de janeiro           62151
belo horizonte           27805
curitiba                 16593
porto alegre             13521
                         ...  
baía formosa                 1
herval doeste                1
cachoeira dos índios         1
joca claudino                1
casserengue                  1
Name: geolocation_city, Length: 8010, dtype: int64


# Work with Cities



In [22]:
customerData_clean= customerData.replace(['são paulo'],['sao paulo'])
customerData_clean["customer_city"].value_counts()



sao paulo            15540
rio de janeiro        6882
belo horizonte        2773
brasilia              2131
curitiba              1521
                     ...  
bequimao                 1
andarai                  1
vargem grande            1
curvelandia              1
eugenio de castro        1
Name: customer_city, Length: 4119, dtype: int64

In [23]:
#Sao Paulo
saoPauloCustomerData = customerData_clean[(customerData_clean == 'sao paulo').any(axis=1)]
customer_idSP= saoPauloCustomerData.iloc[:,0]
customers_idSP_list = customer_idSP.to_list()
type(customers_idSP_list[0])
customer_idSP
#customers_idSP_list


2        4e7b3e00288586ebd08712fdd0374a03
6        fd826e7cf63160e536e0908c76c3f441
13       eabebad39a88bb6f5b52376faec28612
18       9b8ce803689b3562defaad4613ef426f
22       2938121a40a20953c43caa8c98787fcb
                       ...               
99408    f6c6d3e1e20969a5eed982163f959719
99411    da37711b17efd5f2539e8196ab215f04
99422    821a7275a08f32975caceff2e08ea262
99423    c6ece8a5137f3c9c3a3a12302a19a2ac
99436    17ddf5dd5d51696bb3d7c6291687be6f
Name: customer_id, Length: 15540, dtype: object

In [24]:
#Sao Paulo
SP_orders = orders[orders['customer_id'].isin(customers_idSP_list)]
SP_orders.head()

order_id                       customer_id  \
0   e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
11  34513ce0c4fab462a55830c0989c7edb  7711cf624183d843aafe81855097bc37   
13  5ff96c15d0b717ac6ad1f3d77225a350  19402a48fe860416adf93348aba37740   
14  432aaf21d85167c2c86ec9448c4e42cc  3df704f53d3f1d4818840b34ec672a9f   
20  203096f03d82e0dffbc41ebc2e2bcfb7  d2b091571da224a1b36412c18bc3bbfe   

   order_status order_purchase_timestamp    order_approved_at  \
0     delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
11    delivered      2017-07-13 19:58:11  2017-07-13 20:10:08   
13    delivered      2018-07-25 17:44:10  2018-07-25 17:55:14   
14    delivered      2018-03-01 14:14:28  2018-03-01 15:10:47   
20    delivered      2017-09-18 14:31:30  2017-09-19 04:04:09   

   order_delivered_carrier_date order_delivered_customer_date  \
0           2017-10-04 19:55:00           2017-10-10 21:25:13   
11          2017-07-14 18:43:29           2017-07-19 14:04:48   
13          2018-07-26 13:16:00           2018-07-30 15:52:25   
14          2018-03-02 21:09:20           2018-03-12 23:36:26   
20          2017-10-06 17:50:03           2017-10-09 22:23:46   

   order_estimated_delivery_date  
0            2017-10-18 00:00:00  
11           2017-08-08 00:00:00  
13           2018-08-08 00:00:00  
14           2018-03-21 00:00:00  
20           2017-09-28 00:00:00

In [25]:
order_idSP= SP_orders.iloc[:,0]
order_idSP_list = order_idSP.to_list()

In [26]:
SP_order_reviews = order_reviews_clean2[order_reviews_clean2['order_id'].isin(order_idSP_list)]
SP_order_reviews.describe()

NameError: name 'order_reviews_clean2' is not defined

In [0]:
# RIO De Janeiro


#RioCustomerData = customerData[(customerData['customer_city'] == 'rio de janeiro')]
RioCustomerData = customerData_clean[(customerData_clean == 'rio de janeiro').any(axis=1)]
customer_idRio= RioCustomerData.iloc[:,0]
customers_idRio_list = customer_idRio.to_list()

len(customers_idRio_list)


In [0]:


RioOrders = orders[orders['customer_id'].isin(customers_idRio_list)]

Rio_order_id_list = RioOrders.iloc[:,0].to_list()


RioReviews = order_reviews_clean2[order_reviews_clean2['order_id'].isin(Rio_order_id_list)]

RioReviews

In [0]:
#belo horizonte  
beloHorizonteCustomerData = customerData_clean[(customerData_clean == 'belo horizonte').any(axis=1)]
customer_idBH= beloHorizonteCustomerData.iloc[:,0]
customers_idBH_list = customer_idBH.to_list()

In [0]:
BH_orders = orders[orders['customer_id'].isin(customers_idBH_list)]
BH_orders

In [0]:
order_idBH= BH_orders.iloc[:,0]
order_idBH_list = order_idBH.to_list()

In [0]:
BH_order_reviews = order_reviews_clean2[order_reviews_clean2['order_id'].isin(order_idBH_list)]
BH_order_reviews.describe()

In [0]:
brasilia_CustomerData = customerData_clean[(customerData_clean['customer_city'] == 'brasilia')]

customer_id_brasilia= brasilia_CustomerData.iloc[:,0]
customers_idbrasilia_list = customer_id_brasilia.to_list()

len(customers_idbrasilia_list)


In [0]:
BrasiliaOrders = orders[orders['customer_id'].isin(customers_idbrasilia_list)]

Brasilia_order_id_list = BrasiliaOrders.iloc[:,0].to_list()


BrasiliaReviews = order_reviews_clean2[order_reviews_clean2['order_id'].isin(Brasilia_order_id_list)]

print(len(BrasiliaReviews))
len(BrasiliaReviews['review_id'].unique())

In [0]:
#curitiba 
curitibaCustomerData = customerData_clean[(customerData_clean == 'curitiba').any(axis=1)]
customer_idCur= curitibaCustomerData.iloc[:,0]
customer_idCur_list = customer_idCur.to_list()


In [0]:
Cur_orders = orders[orders['customer_id'].isin(customer_idCur_list)]
Cur_orders

In [0]:
order_idCur= Cur_orders.iloc[:,0]
order_idCur_list = order_idCur.to_list()

In [0]:


Cur_order_reviews = order_reviews_clean2[order_reviews_clean2['order_id'].isin(order_idCur_list)]
Cur_order_reviews.describe()
# print(len(Cur_order_reviews))
# len(Cur_order_reviews['review_id'].unique())

In [10]:
#DUPLICATE REVIEWS EVEN THOUGH EACH REVIEW ID SHOULD BE UNIQUE

print(len(order_reviews))
print(len(order_reviews['review_id'].unique()))

order_reviews_clean2 = order_reviews_clean.drop_duplicates(subset=['review_id'])
len(order_reviews_clean2)

99224
98410


NameError: name 'order_reviews_clean' is not defined

In [11]:
order_reviews_clean.isna().sum()

NameError: name 'order_reviews_clean' is not defined

In [12]:
order_w_customer = pd.merge(orders_clean,customerData_clean, on = "customer_id" )
order_items_w_ordercust = pd.merge(order_items_clean, order_w_customer, on = "order_id")

order_items_w_ordercust['customer_unique_id'].value_counts().mean()

order_items_everything = pd.merge(order_items_w_ordercust, products_english[['product_id','product_category_name_english']])

NameError: name 'orders_clean' is not defined

# Orders Clean



In [13]:
nan_orders = orders[orders.isna().any(axis=1)]
orders_listofnans = nan_orders['order_approved_at']
orders_listofnans= orders_listofnans.squeeze()



orders_clean = orders[~orders['order_approved_at'].isin(orders_listofnans)]

orders_clean.shape


# nan_products = products[products.isna().any(axis=1)]
# listofnans = nan_products.iloc[:, :1]
# listofnans= listofnans.squeeze()
# orders.isna().sum()

                      
                    

(96010, 8)

In [14]:
nan_orders

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp    order_approved_at  \
6         invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
44         shipped      2018-06-04 16:44:48  2018-06-05 04:31:18   
103       invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
154        shipped      2018-01-10 11:33:07  2018-01-11 02:32:30   
...            ...                      ...                  ...   
99283     canceled      2018-08-31 16:13:44                  NaN   
99313   processing      2018-01-05 23:07:24  2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                  NaN   
99348  unavailable      2017-08-23 16:28:04  2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03  2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaN                           NaN   
44             2018-06-05 14:32:00                           NaN   
103                            NaN                           NaN   
128                            NaN                           NaN   
154            2018-01-11 19:39:23                           NaN   
...                            ...                           ...   
99283                          NaN                           NaN   
99313                          NaN                           NaN   
99347                          NaN                           NaN   
99348                          NaN                           NaN   
99415                          NaN                           NaN   

      order_estimated_delivery_date  
6               2017-05-09 00:00:00  
44              2018-06-28 00:00:00  
103             2018-08-21 00:00:00  
128             2017-10-03 00:00:00  
154             2018-02-07 00:00:00  
...                             ...  
99283           2018-10-01 00:00:00  
99313           2018-02-06 00:00:00  
99347           2018-09-27 00:00:00  
99348           2017-09-15 00:00:00  
99415           2017-10-23 00:00:00  

[2980 rows x 8 columns]

In [15]:
order_items_everything2 = pd.merge(order_items_everything,sellers,on = "seller_id")

NameError: name 'order_items_everything' is not defined

In [16]:
order_items_everything2['order_delivered_customer_date'] = pd.to_datetime(order_items_everything2['order_delivered_customer_date'])
order_items_everything2['order_approved_at'] = pd.to_datetime(order_items_everything2['order_approved_at'])
order_items_everything2['order_purchase_timestamp'] = pd.to_datetime(order_items_everything2['order_purchase_timestamp'])
order_items_everything2['order_estimated_delivery_date'] = pd.to_datetime(order_items_everything2['order_estimated_delivery_date'])
order_items_everything2['order_delivered_carrier_date'] = pd.to_datetime(order_items_everything2['order_delivered_carrier_date'])


NameError: name 'order_items_everything2' is not defined

In [17]:
order_items_everything2['delivery_time'] = order_items_everything2['order_delivered_customer_date'] -order_items_everything2['order_approved_at']
order_items_everything2['estimated_delivery_time'] = order_items_everything2['order_estimated_delivery_date'] -order_items_everything2['order_approved_at']


NameError: name 'order_items_everything2' is not defined

In [18]:
order_items_slimmed = order_items_everything2.drop(columns = ['freight_value','order_status','order_purchase_timestamp','order_approved_at','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date','customer_zip_code_prefix','seller_zip_code_prefix'])



NameError: name 'order_items_everything2' is not defined

In [19]:
order_items_slimmed['Delivery_time_difference'] = order_items_slimmed['estimated_delivery_time'] - order_items_slimmed['delivery_time']

NameError: name 'order_items_slimmed' is not defined

In [20]:
order_items_slimmed

NameError: name 'order_items_slimmed' is not defined

In [32]:
order_items_everything2["customer_unique_id"].value_counts()


c8460e4251689ba205045f3ea17884a1    24
4546caea018ad8c692964e3382debd19    21
c402f431464c72e27330a67f7b94d4fb    20
698e1cf81d01a3d389d96145f7fa6df8    20
0f5ac8d5c31de21d2f25e24be15bbffb    18
                                    ..
7276dbdcf4c8838b39accfacf236480a     1
198476e6d4b537f6fc71e4f3adaa52d1     1
fdc85558580c10498bbe86fe0fb9d996     1
b634a30cb90c1cbbff45bf9417cff90e     1
7f686e118b3f870f0f28d152f3dfacc6     1
Name: customer_unique_id, Length: 91639, dtype: int64

# LINKING ORDERS WITH CUSTOMERS



In [33]:
orders['order_purchase_timestamp']= pd.to_datetime(orders['order_purchase_timestamp'])
orders['order_approved_at']= pd.to_datetime(orders['order_approved_at'])
orders['order_delivered_carrier_date']= pd.to_datetime(orders['order_delivered_carrier_date'])
orders['order_estimated_delivery_date']= pd.to_datetime(orders['order_estimated_delivery_date'])
orders['order_delivered_customer_date']= pd.to_datetime(orders['order_delivered_customer_date'])


# print(products_clean['product_category_name'].unique())
perfume = products_clean[(products_clean == 'perfumaria').any(axis=1)]
perfume["product_category_name"].value_counts()

perfumaria    868
Name: product_category_name, dtype: int64

In [34]:
products.shape[0]
order_items.shape

orders.shape

#len(order_items['order_item_id'].unique())
#len(orders)


(99441, 8)

In [35]:
order_items['order_item_id'] = pd.to_numeric(order_items['order_item_id'])

order_items[order_items['order_item_id']> 1]

order_id  order_item_id  \
14      0008288aa423d2a3f00fcb17cd7d8719              2   
33      00143d0f86d6fbd9f9b38ab440ac16f5              2   
34      00143d0f86d6fbd9f9b38ab440ac16f5              3   
43      001ab0a7578dd66cd4b0a71f5b6e1e41              2   
44      001ab0a7578dd66cd4b0a71f5b6e1e41              3   
...                                  ...            ...   
112617  ffecd5a79a0084f6a592288c67e3c298              3   
112635  fff8287bbae429a99bb7e8c21d151c41              2   
112641  fffb9224b6fc7c43ebb0904318b10b5f              2   
112642  fffb9224b6fc7c43ebb0904318b10b5f              3   
112643  fffb9224b6fc7c43ebb0904318b10b5f              4   

                              product_id                         seller_id  \
14      368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   
33      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
34      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
43      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
44      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
...                                  ...                               ...   
112617  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112635  bee2e070c39f3dd2f6883a17a5f0da45  4e922959ae960d389249c378d1c939f5   
112641  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112642  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112643  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   

        shipping_limit_date   price  freight_value  
14      2018-02-21 02:55:52   49.90          13.37  
33      2017-10-20 16:07:52   21.33          15.10  
34      2017-10-20 16:07:52   21.33          15.10  
43      2018-01-04 02:33:42   24.89          17.63  
44      2018-01-04 02:33:42   24.89          17.63  
...                     ...     ...            ...  
112617  2018-03-05 20:15:27   21.90          15.79  
112635  2018-03-27 12:29:22  180.00          48.14  
112641  2017-11-03 02:55:58   55.00          34.19  
112642  2017-11-03 02:55:58   55.00          34.19  
112643  2017-11-03 02:55:58   55.00          34.19  

[13984 rows x 7 columns]

In [15]:
delivery_time = orders["order_delivered_customer_date"] - orders['order_approved_at']
delivery_time
delivery_time.mean()



Timedelta('12 days 03:07:42.855694470')

In [63]:
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [17]:
len(geolocation['geolocation_zip_code_prefix'].unique())

19015

In [0]:
orders_w_customers = pd.merge(customerData_clean)

In [43]:
order_items_clean["order_item_id"].max()

21

In [17]:
est_delivery_time = orders["order_estimated_delivery_date"] - orders['order_approved_at']


est_delivery_time.mean()

Timedelta('23 days 08:01:22.510722092')

In [0]:
delivery_time = orders["order_delivered_customer_date"] - orders['order_estimated_delivery_date']

In [13]:
orders['order_delivered_customer_date']
   
# print(type(orders.iloc[4]))

Timestamp('2018-02-02 19:28:10.500000')

**Avg review score      \-         4.08642062404257**

**Number of unique customers       \-       96096**

**Total orders      \-       99441**

**Average orders per customer        \-        1.0348089410589412**

**Average items ordered per customer \-  1.1978339991122948**


In [153]:
#             CUSTOMER DATA              #
customerData.isna().sum()
customerData.head()
print(len(customerData))

customerData['customer_unique_id'].value_counts()


99441


1.0348089410589412

In [109]:
#             GEOLOCATION              #
geolocation.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [26]:
#             PAYMENTS              #
payments.isna().sum().sum

<bound method NDFrame._add_numeric_operations.<locals>.sum of order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64>

In [11]:
#             PRODUCTS              #
nan_products = products[products.isna().any(axis=1)]
listofnans = nan_products.iloc[:, :1]
listofnans= listofnans.squeeze()

order_items_clean = order_items

# for i in order_items['product_id']:
#     for j in listofnans:
#         if (i==j):
#             order_items_clean.drop(i)
# print(len(order_items))""
# print(len(order_items_clean))

products.isna().sum().sum()
products_clean = products.dropna()

nan_products.iloc[:, :1]

product_id
105    a41e356c76fab66334f36de622ecbd3a
128    d8dee61c2034d6d075997acef1870e9b
145    56139431d72cd51f19eb9f7dae4d1617
154    46b48281eb6d663ced748f324108c733
197    5fb61f482620cb672f5e586bb132eae9
...                                 ...
32515  b0a0c5dd78e644373b199380612c350a
32589  10dbe0fbaa2c505123c17fdc34a63c56
32616  bd2ada37b58ae94cc838b9c0569fecd8
32772  fa51e914046aab32764c41356b9d4ea4
32852  c4ceee876c82b8328e9c293fa0e1989b

[611 rows x 1 columns]

In [4]:
listofnans

105      a41e356c76fab66334f36de622ecbd3a
128      d8dee61c2034d6d075997acef1870e9b
145      56139431d72cd51f19eb9f7dae4d1617
154      46b48281eb6d663ced748f324108c733
197      5fb61f482620cb672f5e586bb132eae9
                       ...               
32515    b0a0c5dd78e644373b199380612c350a
32589    10dbe0fbaa2c505123c17fdc34a63c56
32616    bd2ada37b58ae94cc838b9c0569fecd8
32772    fa51e914046aab32764c41356b9d4ea4
32852    c4ceee876c82b8328e9c293fa0e1989b
Name: product_id, Length: 611, dtype: object

In [12]:
order_items_clean = order_items[~order_items['product_id'].isin(listofnans)]

print(len(order_items_clean))
print(len(order_items))

# order_items['product_id'].isunique

# ORDER ITEMS CLEAN GOOD
# PRODUCTS CLEAN GOOD

111046
112650


# ORDER\_ITEMS\_CLEAN



In [89]:
# TESTING TO SEE IF we cleaned
list_of_lengths = []
list_of_lengths_clean = []


for col in order_items:
    length = len(order_items[col].unique())
    list_of_lengths.append(length)
print(list_of_lengths[2])

for col in order_items_clean:
    list_of_lengths_clean.append(len(order_items_clean[col].unique()))
print(list_of_lengths_clean[2])
# print(type(list_of_order_items_clean))


32951


32340


In [54]:

for i in order_items['product_id']:
    for j in listofnans:
        print(j)
        if (i==j):
            print(i)

In [58]:
listofnans.squeeze()

105      a41e356c76fab66334f36de622ecbd3a
128      d8dee61c2034d6d075997acef1870e9b
145      56139431d72cd51f19eb9f7dae4d1617
154      46b48281eb6d663ced748f324108c733
197      5fb61f482620cb672f5e586bb132eae9
                       ...               
32515    b0a0c5dd78e644373b199380612c350a
32589    10dbe0fbaa2c505123c17fdc34a63c56
32616    bd2ada37b58ae94cc838b9c0569fecd8
32772    fa51e914046aab32764c41356b9d4ea4
32852    c4ceee876c82b8328e9c293fa0e1989b
Name: product_id, Length: 611, dtype: object

In [21]:
#             PRODUCT CATEGORY NAME              #
nan_prod_cat_name = prod_cat[prod_cat.isna().any(axis=1)]
prod_cat_name_listofnans = nan_prod_cat_name.iloc[:, :1]
prod_cat_name_listofnans = prod_cat_name_listofnans.squeeze()

prod_cat_name_clean = prod_cat[~prod_cat['product_category_name'].isin(prod_cat_name_listofnans)]



prod_cat_name_clean['product_category_name_english']

0                 health_beauty
1         computers_accessories
2                          auto
3                bed_bath_table
4               furniture_decor
                ...            
66                      flowers
67        arts_and_craftmanship
68          diapers_and_hygiene
69    fashion_childrens_clothes
70        security_and_services
Name: product_category_name_english, Length: 71, dtype: object

In [131]:
prod_cat_name_clean.shape

(71, 2)

In [104]:
#             ORDERS              #

# print(orders.isna().sum())
# nan_orders = orders[orders.isna().any(axis=1)]
orders_listofnans = nan_orders
orders_listofnans= orders_listofnans.squeeze()

# order_items_clean = order_items


# for i in order_reviews['order_id']:
#     for j in orders_listofnans:
#         if (i==j):
#             orders_clean.drop(i)


# print(len(orders))
# print(len(orders_clean))


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


99441
99441


In [119]:
#             SELLERS              #
print(sellers.isna().sum())
sellers.head()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64


seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [107]:
payments.isna().sum()


order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [110]:
order_reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

# Order Reviews



In [8]:
order_reviews_clean = order_reviews.drop(['review_comment_title','review_comment_message'],axis=1)
order_reviews_clean


review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_creation_date review_answer_timestamp  
0                 4  2018-01-18 00:00:00     2018-01-18 21:46:59  
1                 5  2018-03-10 00:00:00     2018-03-11 03:05:13  
2                 5  2018-02-17 00:00:00     2018-02-18 14:36:24  
3                 5  2017-04-21 00:00:00     2017-04-21 22:02:06  
4                 5  2018-03-01 00:00:00     2018-03-02 10:26:53  
...             ...                  ...                     ...  
99219             5  2018-07-07 00:00:00     2018-07-14 17:18:30  
99220             5  2017-12-09 00:00:00     2017-12-11 20:06:42  
99221             5  2018-03-22 00:00:00     2018-03-23 09:10:43  
99222             4  2018-07-01 00:00:00     2018-07-02 12:59:13  
99223             1  2017-07-03 00:00:00     2017-07-03 21:01:49  

[99224 rows x 5 columns]

In [132]:
products_clean["product_category_name"].value_counts()


cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: product_category_name, Length: 73, dtype: int64

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm product_category_name_english  
0                  14.0                 health_beauty  
1                  20.0         computers_accessories  
2                  15.0                          auto  
3                  26.0                bed_bath_table  
4                  13.0               furniture_decor  
...                 ...                           ...  
32946              40.0                           NaN  
32947              16.0                           NaN  
32948              27.0                           NaN  
32949              20.0                           NaN  
32950               7.0                           NaN  

[32340 rows x 10 columns]

In [125]:
products

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32951 rows x 9 columns]

In [52]:
print(order_reviews_clean['review_score'].mean())

NameError: name 'order_reviews_clean' is not defined

In [121]:
prod_cat.isna().sum()



product_category_name            0
product_category_name_english    0
dtype: int64

In [123]:
order

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [151]:
orders

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [8]:
#Highest rated sellers
sellers
sellers_count= sellers['seller_id'].value_counts()
sellers_count
order_items

3442f8959a84dea7ee197c632cb2df15    1
e26901d5ab434ce92fd9b5c256820a4e    1
7e3f87d16fb353f408d467e74fbd8014    1
9140bd1d9de28b197dce077bdd687438    1
a2b911e9ad40dc7d61c142a4683e2d85    1
                                   ..
f918546be4bec7104d117789ba758b2b    1
eb4a59a06b3948e851a7d7a830bcbc42    1
80146bc52c4f6af5239b23de2a20262b    1
4091f77d8c768fc3e81c20761471761d    1
9e25199f6ef7e7c347120ff175652c3b    1
Name: seller_id, Length: 3095, dtype: int64

In [106]:
top_sellers =order_items['seller_id'].value_counts()
sellers_200 = top_sellers[top_sellers >200]

sellers_200.index[0]
#list1.mean()
#order_items.groupby('seller_id')['order_id'].count()

6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
                                    ... 
400f221ab83037c0ea78f1267da9624d       1
1746c1ae87ac63d530c7c943d0ac42e2       1
16bdc8cefd0e32a6f0824d296c5ad14a       1
ace601496e211a664100321f038f50af       1
13d95f0f6f73943d4ceffad0fc2cd32c       1
Name: seller_id, Length: 3095, dtype: int64


'6560211a19b47992c3666cc44a7e94c0'

In [42]:
top_seller_ids = sellers_200.index

In [82]:
top_seller_orders = order_items[order_items['seller_id'].isin(top_seller_ids)]

# print(top_seller_orders)
top_seller_order_ids = top_seller_orders.iloc[:,0:]

sellers_orders_ids = top_seller_orders[["order_id","seller_id"]]
print(sellers_orders_ids)
print(len(sellers_orders_ids['order_id'].unique()))
len(sellers_orders_ids['seller_id'].unique())
# print(top_seller_order_ids)

                                order_id                         seller_id
6       00054e8431b9d7675808bcb819fb4a32  7040e82f899a04d1b434b795a43b4617
10      00061f2a7bc09da83e415a52dc8a4af1  cc419e0650a3c5ba77189a1882b7556a
12      0006ec9db01a64e59a68b2c340bf65a7  4a3ca9315b744ce9f8e9374361493884
13      0008288aa423d2a3f00fcb17cd7d8719  1f50f920176fa81dab994f9023523100
14      0008288aa423d2a3f00fcb17cd7d8719  1f50f920176fa81dab994f9023523100
...                                  ...                               ...
112632  fff7c4452f050315db1b3f24d9df5fcd  7e93a43ef30c4f03f38b393420bc753a
112633  fff8286f77788ab8b55b2e5747fa7dd8  ea8482cd71df3c1969d7b9473ff13abc
112634  fff8287bbae429a99bb7e8c21d151c41  4e922959ae960d389249c378d1c939f5
112635  fff8287bbae429a99bb7e8c21d151c41  4e922959ae960d389249c378d1c939f5
112647  fffce4705a9662cd70adb13d4a31832d  c3cfdc648177fdbbbb35635a37472c53

[51026 rows x 2 columns]
45172


102

In [92]:
reviews_with_sellers= pd.merge(order_reviews_clean2, sellers_orders_ids, on='order_id')

reviews_with_sellers

review_id                          order_id  \
0      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
1      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
2      15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
3      15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
4      15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
...                                 ...                               ...   
50530  a709d176f59bc3af77f4149c96bae357  d5cb12269711bd1eaf7eed8fd32a7c95   
50531  115d72425fc59eec9a2e9d7594c6d93d  acd45245723df7cb52772a34416b41b1   
50532  c6b270c61f67c9f7cb07d84ea8aeaf8b  48f7ee67313eda32bfcf5b9c1dd9522d   
50533  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
50534  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   

       review_score review_creation_date review_answer_timestamp  \
0                 5  2018-02-17 00:00:00     2018-02-18 14:36:24   
1                 5  2017-04-21 00:00:00     2017-04-21 22:02:06   
2                 1  2018-04-13 00:00:00     2018-04-16 00:39:37   
3                 1  2018-04-13 00:00:00     2018-04-16 00:39:37   
4                 1  2018-04-13 00:00:00     2018-04-16 00:39:37   
...             ...                  ...                     ...   
50530             3  2018-05-19 00:00:00     2018-05-20 21:51:06   
50531             3  2018-05-08 00:00:00     2018-05-10 14:37:03   
50532             5  2017-12-13 00:00:00     2017-12-14 11:09:36   
50533             5  2017-12-09 00:00:00     2017-12-11 20:06:42   
50534             4  2018-07-01 00:00:00     2018-07-02 12:59:13   

                              seller_id  
0      a1043bafd471dff536d0c462352beb48  
1      a1043bafd471dff536d0c462352beb48  
2      da8622b14eb17ae2831f4ac5b9dab84a  
3      da8622b14eb17ae2831f4ac5b9dab84a  
4      da8622b14eb17ae2831f4ac5b9dab84a  
...                                 ...  
50530  562fc2f2c2863ab7e79a9e4388a58a14  
50531  dbc22125167c298ef99da25668e1011f  
50532  6560211a19b47992c3666cc44a7e94c0  
50533  46dc3b2cc0980fb8ec44634e21d2718e  
50534  8d956fec2e4337affcb520f56fd8cbfd  

[50535 rows x 6 columns]

In [116]:
mean_review_top200 = reviews_with_sellers.groupby("seller_id")['review_score'].mean()


seller_id
001cca7ae9ae17fb1caed9dfb1094831    3.902542
0241d4d5d36f10f80c644447315af0bd    4.297959
06a2c3af7b3aee5d69171b0e14f0ee87    4.002506
0b90b6df587eb83608a64ea8b390cf07    4.023585
0c8380b62e38e8a1e6adbeba7eb9688c    4.041284
                                      ...   
f7ba60f8c3f99e7ee4042fdef03b70c4    4.207792
f8db351d8c4c4c22c6835c19a46f01b0    4.216179
fa1c13f2614d7b5c4749cbc52fecda94    4.340206
fa40cc5b934574b62717c68f3d678b6d    4.430678
fe2032dab1a61af8794248c8196565c9    4.395349
Name: review_score, Length: 102, dtype: float64

In [128]:
regression_table = pd.concat([mean_review_top200, sellers_200], axis=1)

regression_table['order_items'] = regression_table['seller_id']

regression_table.drop('seller_id',axis=1, inplace= True)
regression_table


review_score  order_items
001cca7ae9ae17fb1caed9dfb1094831      3.902542          239
0241d4d5d36f10f80c644447315af0bd      4.297959          246
06a2c3af7b3aee5d69171b0e14f0ee87      4.002506          405
0b90b6df587eb83608a64ea8b390cf07      4.023585          210
0c8380b62e38e8a1e6adbeba7eb9688c      4.041284          220
...                                        ...          ...
f7ba60f8c3f99e7ee4042fdef03b70c4      4.207792          230
f8db351d8c4c4c22c6835c19a46f01b0      4.216179          724
fa1c13f2614d7b5c4749cbc52fecda94      4.340206          586
fa40cc5b934574b62717c68f3d678b6d      4.430678          340
fe2032dab1a61af8794248c8196565c9      4.395349          306

[102 rows x 2 columns]

In [121]:
top_sellers

6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
                                    ... 
400f221ab83037c0ea78f1267da9624d       1
1746c1ae87ac63d530c7c943d0ac42e2       1
16bdc8cefd0e32a6f0824d296c5ad14a       1
ace601496e211a664100321f038f50af       1
13d95f0f6f73943d4ceffad0fc2cd32c       1
Name: seller_id, Length: 3095, dtype: int64

In [100]:
top_seller_reviews = order_reviews_clean2[order_reviews_clean2['order_id'].isin(top_seller_order_ids)]

# top_seller_reviews.dtypes

#combined_table = top_seller_reviews.join(top_seller_orders, on = "order_id")


In [156]:
customerData['customer_unique_id'].value_counts()

8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
ca77025e7201e3b30c44b472ff346268     7
6469f99c1f9dfae7733b25662e7f1782     7
                                    ..
081f07439678af2da7755a0aa572154d     1
9c57789a6a587f4efcfb01e36c106014     1
9e4c21635f640562de1a6976374fba03     1
06c0ea6ee892364d1608ee47aa9f56a0     1
84732c5050c01db9b23e19ba39899398     1
Name: customer_unique_id, Length: 96096, dtype: int64

In [18]:
customerData_value_counts = customerData['customer_unique_id'].value_counts()
type(customerData_value_counts)
customerData_value_counts.value_counts()

1     93099
2      2745
3       203
4        30
5         8
6         6
7         3
17        1
9         1
Name: customer_unique_id, dtype: int64

In [164]:
customer_ids = customerData[['customer_id','customer_unique_id']]

In [165]:
orders1 = pd.merge(orders,customer_ids, on = "customer_id")

In [166]:
orders1

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
0        delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...            ...                      ...                  ...   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date                customer_unique_id  
0               2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff  
1               2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231  
2               2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8  
3               2017-12-15 00:00:00  7c142cf63193a1473d2e66489a9ae977  
4               2018-02-26 00:00:00  72632f0f9dd73dfee390c9b22eb56dd6  
...                             ...                               ...  
99436           2017-03-28 00:00:00  6359f309b166b0196dbf7ad2ac62bb5a  
99437           2018-03-02 00:00:00  da62f9e57a76d978d02ab5362c509660  
99438           2017-09-27 00:00:00  737520a9aad80b3fbbdad19b66b37b30  
99439           2018-02-15 00:00:00  5097a5312c8b157bb7be58ae360ef43c  
99440           2018-04-03 00:00:00  60350aa974b26ff12caad89e55993bd6  

[99441 rows x 9 columns]

In [169]:
order_reviews_clean

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_creation_date review_answer_timestamp  
0                 4  2018-01-18 00:00:00     2018-01-18 21:46:59  
1                 5  2018-03-10 00:00:00     2018-03-11 03:05:13  
2                 5  2018-02-17 00:00:00     2018-02-18 14:36:24  
3                 5  2017-04-21 00:00:00     2017-04-21 22:02:06  
4                 5  2018-03-01 00:00:00     2018-03-02 10:26:53  
...             ...                  ...                     ...  
99219             5  2018-07-07 00:00:00     2018-07-14 17:18:30  
99220             5  2017-12-09 00:00:00     2017-12-11 20:06:42  
99221             5  2018-03-22 00:00:00     2018-03-23 09:10:43  
99222             4  2018-07-01 00:00:00     2018-07-02 12:59:13  
99223             1  2017-07-03 00:00:00     2017-07-03 21:01:49  

[99224 rows x 5 columns]

In [170]:
unique_reviews = pd.merge(order_reviews_clean,orders1[['order_id','customer_unique_id']], on = "order_id")

In [177]:
unique_reviews
unique_reviews["customer_unique_id"].value_counts()

8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
ca77025e7201e3b30c44b472ff346268     7
b4e4f24de1e8725b74e4a1f4975116ed     7
                                    ..
9635cc2dc37d7d02dea728765b4055f2     1
2e65fa6556bc340335534ea848f385ff     1
af21f4525831bb14040012fd9bfec6d3     1
8cfecfbbdad82f1281686a54caf46dbb     1
432d4f5ec4f462779a48a51c33950351     1
Name: customer_unique_id, Length: 95380, dtype: int64

In [178]:
prod_cat

product_category_name product_category_name_english
0                    beleza_saude                 health_beauty
1          informatica_acessorios         computers_accessories
2                      automotivo                          auto
3                 cama_mesa_banho                bed_bath_table
4                moveis_decoracao               furniture_decor
..                            ...                           ...
66                         flores                       flowers
67             artes_e_artesanato         arts_and_craftmanship
68                fraldas_higiene           diapers_and_hygiene
69  fashion_roupa_infanto_juvenil     fashion_childrens_clothes
70             seguros_e_servicos         security_and_services

[71 rows x 2 columns]

In [23]:
products_english = pd.merge(products_clean,prod_cat, on  = "product_category_name")
products_english
#products_english[products_english['product_category_name_english']== "cool_stuff"]

product_id      product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                 perfumaria   
1      6a2fb4dd53d2cdb88e0432f1284a004c                 perfumaria   
2      0d009643171aee696f4733340bc2fdd0                 perfumaria   
3      b1eae565a61935e0011ee7682fef9dc9                 perfumaria   
4      8da90b37f0fb171b4877c124f965b1f6                 perfumaria   
...                                 ...                        ...   
32322  13b25797ad1e6d6c7cd3cbeb35f8ee7a  moveis_colchao_e_estofado   
32323  dc36a7859b743d8610a2bbbaea26ece9  moveis_colchao_e_estofado   
32324  107fde0930956120d1e13dd1062fbb46  moveis_colchao_e_estofado   
32325  726b4e18f00255e2e63491bcba3f60b8  moveis_colchao_e_estofado   
32326  1dceebcc5f23c02ea23e16d5bedca000          cds_dvds_musicais   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     39.0                       346.0                 2.0   
2                     52.0                       150.0                 1.0   
3                     49.0                       460.0                 2.0   
4                     56.0                       733.0                 3.0   
...                    ...                         ...                 ...   
32322                 44.0                       348.0                 1.0   
32323                 29.0                      2073.0                 2.0   
32324                 51.0                       418.0                 1.0   
32325                 41.0                       866.0                 1.0   
32326                 47.0                       117.0                 2.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                 400.0               27.0                5.0   
2                 422.0               21.0               16.0   
3                 267.0               17.0               13.0   
4                 377.0               18.0               13.0   
...                 ...                ...                ...   
32322           18250.0               25.0               25.0   
32323            2500.0               41.0               21.0   
32324           30000.0               60.0               73.0   
32325            1700.0               70.0               10.0   
32326             550.0               35.0               15.0   

       product_width_cm      product_category_name_english  
0                  14.0                          perfumery  
1                  20.0                          perfumery  
2                  18.0                          perfumery  
3                  17.0                          perfumery  
4                  15.0                          perfumery  
...                 ...                                ...  
32322              25.0  furniture_mattress_and_upholstery  
32323              21.0  furniture_mattress_and_upholstery  
32324              60.0  furniture_mattress_and_upholstery  
32325             100.0  furniture_mattress_and_upholstery  
32326              25.0                  cds_dvds_musicals  

[32327 rows x 10 columns]

In [181]:
order_items_w_cat = pd.merge(order_items_clean, products_english[['product_id','product_category_name_english']], on = "product_id")

In [182]:
order_items_w_cat

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       130898c0987d1801452a8ed92a670612              1   
2       532ed5e14e24ae1f0d735b91524b98b9              1   
3       6f8c31653edb8c83e1a739408b5ff750              1   
4       7d19f4ef4d04461989632411b7e588b9              1   
...                                  ...            ...   
111017  ffebd80e3291e811c308365936897efd              1   
111018  ffee31fb4b5e35c9123608015637c495              1   
111019  fff7c4452f050315db1b3f24d9df5fcd              1   
111020  fffa82886406ccf10c7b4e35c4ff2788              1   
111021  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
2       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
3       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
4       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
...                                  ...                               ...   
111017  4cc4d02efc8f249c13355147fb44e34d  d1c7fa84e48cfa21a0e595167c1c500e   
111018  b10ecf8e33aaaea419a9fa860ea80fb5  0241d4d5d36f10f80c644447315af0bd   
111019  dd469c03ad67e201bc2179ef077dcd48  7e93a43ef30c4f03f38b393420bc753a   
111020  bbe7651fef80287a816ead73f065fc4b  8f2ce03f928b567e3d56181ae20ae952   
111021  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date  price  freight_value  \
0       2017-09-19 09:45:35   58.9          13.29   
1       2017-07-05 02:44:11   55.9          17.96   
2       2018-05-23 10:56:25   64.9          18.33   
3       2017-08-07 18:55:08   58.9          16.17   
4       2017-08-16 22:05:11   58.9          13.29   
...                     ...    ...            ...   
111017  2018-07-18 20:23:55  129.9          51.20   
111018  2018-08-21 10:10:11   99.0          13.52   
111019  2017-06-07 17:05:23  736.0          20.91   
111020  2017-12-22 17:31:42  229.9          44.02   
111021  2018-06-12 17:10:13   43.0          12.79   

       product_category_name_english  
0                         cool_stuff  
1                         cool_stuff  
2                         cool_stuff  
3                         cool_stuff  
4                         cool_stuff  
...                              ...  
111017                  garden_tools  
111018               furniture_decor  
111019                 watches_gifts  
111020                sports_leisure  
111021                bed_bath_table  

[111022 rows x 8 columns]

In [184]:
product_cat_reviews = pd.merge(unique_reviews, order_items_w_cat[['order_id','product_id','product_category_name_english']], on = 'order_id')

In [189]:
print(product_cat_reviews['product_id'].value_counts())
print(product_cat_reviews['product_category_name_english'].value_counts())

aca2eb7d00ea1a7b8ebd4e68314663af    524
422879e10f46682990de24d770e7f83d    486
99a4788cb24856965c36a24e339b6058    482
389d119b48cf3043d311335e499d9c6b    391
368c6c730842d78016ad823897a372db    388
                                   ... 
a12ece3f747b6aae747ca17176611aa2      1
0ad10ec3dc489499e5d7cd19b5181ffe      1
3b85e44a54ca23950c469c7fb1cc0e0a      1
b4be8f8fe79a6219017a8de82c24a4ef      1
4fb209073dc2ab6a6d3c259a7c285501      1
Name: product_id, Length: 32170, dtype: int64
bed_bath_table               11137
health_beauty                 9645
sports_leisure                8640
furniture_decor               8331
computers_accessories         7849
                             ...  
arts_and_craftmanship           24
cds_dvds_musicals               14
la_cuisine                      13
fashion_childrens_clothes        8
security_and_services            2
Name: product_category_name_english, Length: 71, dtype: int64


In [25]:
order_items_slimmed


order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       130898c0987d1801452a8ed92a670612              1   
2       532ed5e14e24ae1f0d735b91524b98b9              1   
3       6f8c31653edb8c83e1a739408b5ff750              1   
4       7d19f4ef4d04461989632411b7e588b9              1   
...                                  ...            ...   
111017  fdaf98feac227eb978d8f33f36e0231d              1   
111018  fe68b52db13993f58175fa589125d345              1   
111019  fefacc66af859508bf1a7934eab1e97f              1   
111020  ff701a7c869ad21de22a6994237c8a00              1   
111021  ff701a7c869ad21de22a6994237c8a00              2   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
2       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
3       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
4       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
...                                  ...                               ...   
111017  f739e9151702508b18f796c53005e5e9  d1aa1ec0839dcab73a6161130eb1f94a   
111018  f4135cbdece8245560f7be179533797a  3fefda3299e6dfaea3466ef346a3571a   
111019  69c590f7ffc7bf8db97190b6cb6ed62e  80ceebb4ee9b31afb6c6a916a574a1e2   
111020  5ff4076c0f01eeba4f728c9e3fa2653c  3e35a8bb43569389d3cebef0ce820f69   
111021  5ff4076c0f01eeba4f728c9e3fa2653c  3e35a8bb43569389d3cebef0ce820f69   

        shipping_limit_date   price                       customer_id  \
0       2017-09-19 09:45:35    58.9  3ce436f183e68e07877b285a838db11a   
1       2017-07-05 02:44:11    55.9  e6eecc5a77de221464d1c4eaff0a9b64   
2       2018-05-23 10:56:25    64.9  4ef55bf80f711b372afebcb7c715344a   
3       2017-08-07 18:55:08    58.9  30407a72ad8b3f4df4d15369126b20c9   
4       2017-08-16 22:05:11    58.9  91a792fef70ecd8cc69d3c7feb3d12da   
...                     ...     ...                               ...   
111017  2017-04-27 05:10:28   119.9  81ae83ce1a10af9f93fa3512bd9713b1   
111018  2017-08-17 02:55:27   209.0  94203a446c1b46b5ac95b0aaf7221804   
111019  2018-08-02 04:05:13  6729.0  f48d464a0baaea338cb25f816991ab1f   
111020  2018-04-18 20:10:33    27.9  8f6d16c31abb30b7b4650272cb0014a3   
111021  2018-04-18 20:10:33    27.9  8f6d16c31abb30b7b4650272cb0014a3   

                      customer_unique_id customer_state  \
0       871766c5855e863f6eccc05f988b23cb             RJ   
1       0fb8e3eab2d3e79d92bb3fffbb97f188             GO   
2       3419052c8c6b45daf79c1e426f9e9bcb             MG   
3       e7c828d22c0682c1565252deefbe334d             PR   
4       0bb98ba72dcc08e95f9d8cc434e9a2cc             MG   
...                                  ...            ...   
111017  3d50a44231c2a153219cef76ee3e445d             RJ   
111018  52a953541861bbb4b79b39df24e00243             SP   
111019  459bef486812aa25204be022145caa62             ES   
111020  ddb2249e0316d365ceae561c0f011bce             SP   
111021  ddb2249e0316d365ceae561c0f011bce             SP   

       product_category_name_english seller_state    delivery_time  \
0                         cool_stuff           SP  7 days 13:58:13   
1                         cool_stuff           SP 14 days 17:55:18   
2                         cool_stuff           SP 17 days 06:02:43   
3                         cool_stuff           SP  8 days 02:31:25   
4                         cool_stuff           SP 13 days 21:59:10   
...                              ...          ...              ...   
111017               furniture_decor           PR 33 days 05:42:15   
111018                   electronics           PR  5 days 14:52:28   
111019                     computers           PR 19 days 10:52:37   
111020                        drinks           SP  7 days 21:38:06   
111021                        

In [86]:
def akinator_1(unique_id):
    r = order_items_slimmed[order_items_slimmed['customer_unique_id'] == unique_id]
    prod_cat_r = r['product_category_name_english'].iloc[0]
    city = r['customer_city'].iloc[0]
    order_items_sub = order_items_slimmed[(order_items_slimmed['product_category_name_english'] == prod_cat_r) & (order_items_slimmed['customer_city'] == city)]
    if len(order_items_sub['seller_id'].unique()) < 10:
        state = r['customer_state'].iloc[0]
        order_items_sub = order_items_slimmed[(order_items_slimmed['product_category_name_english'] == prod_cat_r) & (order_items_slimmed['customer_state'] == state)]
    sorted_order_items_sub = order_items_sub.sort_values(by = ['delivery_time'])
    rec_list_prod_id = sorted_order_items_sub.iloc[0:3,:].sort_values(by = ['price'])['product_id']
    rec_list_prod_id = rec_list_prod_id.tolist()
    rec_list_price = sorted_order_items_sub.iloc[0:3,:].sort_values(by = ['price'])['price']
    rec_list_price = rec_list_price.tolist()
    zipped_rec = zip(rec_list_prod_id, rec_list_price)
    zipped_rec_list = list(zipped_rec)
    new = '\n'
    recs = print(
        f"Based on your previous order history and location, we reccomend:{new}"
        f"Product ID: {zipped_rec_list[0][0]} with price: ${zipped_rec_list[0][1]}{new}"
        f"Product ID: {zipped_rec_list[1][0]} with price: ${zipped_rec_list[1][1]}{new}"
        f"Product ID: {zipped_rec_list[2][0]} with price: ${zipped_rec_list[2][1]}"
    )
    return recs



In [84]:
#def akinator_multi

In [85]:
ok = akinator_1("e7c828d22c0682c1565252deefbe334d")

ok

Based on your previous order history and location, we reccomend:
Product ID: 74e48c5377480b7d048d7b294abc569b with price: $75.0.
Product ID: 3354a4e684f5e7199f9407db70ccd92b with price: $89.99.
Product ID: 165f86fe8b799a708a20ee4ba125c289 with price: $169.99.


In [92]:
user = "c402f431464c72e27330a67f7b94d4fb"
loc = order_items_slimmed['customer_unique_id'].value_counts()
print(loc[loc.index == user])

len(order_items_slimmed[order_items_slimmed['customer_unique_id']== user])

c402f431464c72e27330a67f7b94d4fb    20
Name: customer_unique_id, dtype: int64


20

In [119]:
def akinator_multi(unique_id):
    r = order_items_slimmed[order_items_slimmed['customer_unique_id'] == unique_id]
    prod_cat_r = r['product_category_name_english']
    unique_cat = []
    for n in prod_cat_r:
        if n in unique_cat:
            continue
        else:
            unique_cat.append(n)
    city = r['customer_city'].iloc[0]
    order_items_sub = order_items_slimmed[(order_items_slimmed['product_category_name_english'].isin(unique_cat)) & (order_items_slimmed['customer_city'] == city)]
    if len(order_items_sub['seller_id'].unique()) < 10:
        state = r['customer_state'].iloc[0]
        order_items_sub = order_items_slimmed[(order_items_slimmed['product_category_name_english'].isin(unique_cat)) & (order_items_slimmed['customer_state'] == state)]
    rec_list_prod_id = sorted_order_items_sub.iloc[0:3,:].sort_values(by = ['price'])['product_id']
    rec_list_prod_id = rec_list_prod_id.tolist()
    rec_list_price = sorted_order_items_sub.iloc[0:3,:].sort_values(by = ['price'])['price']
    rec_list_price = rec_list_price.tolist()
    zipped_rec = zip(rec_list_prod_id, rec_list_price)
    zipped_rec_list = list(zipped_rec)
    recs = []
    for q in unique_cat:
        cat_table = order_items_sub[order_items_sub['product_category_name_english']==q].sort_values(by = ['delivery_time'])
        rec_list_prod_id = cat_table.iloc[0:3].sort_values(by = ['price'])['product_id']
        rec_list_price= cat_table.iloc[0:3].sort_values(by = ['price'])['price']
        rec_list_prod_id= list(rec_list_prod_id)
        rec_list_price= list(rec_list_price)
        zipped_rec = zip(rec_list_prod_id, rec_list_price)
        zipped_rec_list = list(zipped_rec)
        new = '\n'
        print(
        f"Based on your previous order history and location, we reccomend:\n"
        f"For {q}:{new}Product ID: {zipped_rec_list[0][0]} with price: ${zipped_rec_list[0][1]}{new}"
        f"Product ID: {zipped_rec_list[1][0]} with price: ${zipped_rec_list[1][1]}{new}"
        f"Product ID: {zipped_rec_list[2][0]} with price: ${zipped_rec_list[2][1]}"
    )

 

In [120]:
akinator_multi("c8460e4251689ba205045f3ea17884a1")

Based on your previous order history and location, we reccomend:
For telephony:
Product ID: 010a3a47c14f1295796c08419f32cef6 with price: $9.99
Product ID: 97017430754804328eb9597b7f85da03 with price: $27.99
Product ID: 430d215740da7fdda91cc938f8fd92f6 with price: $104.0


In [121]:
def akinator(unique_id):
    if len(order_items_slimmed[order_items_slimmed['customer_unique_id']== unique_id]) > 1:
        print("used multi")
        return akinator_multi(unique_id)
    else:
        return akinator_1(unique_id)

In [124]:
akinator("0fb8e3eab2d3e79d92bb3fffbb97f188")

Based on your previous order history and location, we reccomend:
Product ID: 057df564fda79ad8fa5e1928dfbe3e8c with price: $11.99
Product ID: 057df564fda79ad8fa5e1928dfbe3e8c with price: $11.99
Product ID: 057df564fda79ad8fa5e1928dfbe3e8c with price: $11.99
